crates/matrix-sdk-ui/src/room_list_service/filters/invite.rs

```diff
@@ -0,0 +1,84 @@
+use matrix_sdk::{Client, RoomListEntry};
+use matrix_sdk_base::RoomState;
+
+use super::Filter;
+
+struct InviteRoomMatcher<F>
+where
+    F: Fn(&RoomListEntry) -> Option<bool>,
+{
+    is_invite: F,
+}
+
+impl<F> InviteRoomMatcher<F>
+where
+    F: Fn(&RoomListEntry) -> Option<bool>,
+{
+    fn matches(&self, room_list_entry: &RoomListEntry) -> bool {
+        if !matches!(room_list_entry, RoomListEntry::Filled(_) | RoomListEntry::Invalidated(_)) {
+            return false;
+        }
+
+        (self.is_invite)(room_list_entry).unwrap_or(false)
+    }
+}
+
+/// Create a new filter that will accept all filled or invalidated entries, but
+/// filters out rooms that are not invites (see
+/// [`matrix_sdk_base::RoomState::Invited`]).
+pub fn new_filter(client: &Client) -> impl Filter {
+    let client = client.clone();
+
+    let matcher = InviteRoomMatcher {
+        is_invite: move |room| match room {
+            RoomListEntry::Filled(room_id) | RoomListEntry::Invalidated(room_id) => {
+                let room = client.get_room(room_id)?;
+                Some(room.state() == RoomState::Invited)
+            }
+            _ => None,
+        },
+    };
+
+    move |room_list_entry| -> bool { matcher.matches(room_list_entry) }
+}
```

Comment:
To be consistent with the `non_left` filter, I think we should revisit the code a little bit:

```suggestion
struct InviteRoomMatcher<F>
where
    F: Fn(&RoomListEntry) -> Option<RoomState>,
{
    state: F,
}

impl<F> InviteRoomMatcher<F>
where
    F: Fn(&RoomListEntry) -> Option<RoomState>,
{
    fn matches(&self, room: &RoomListEntry) -> bool {
        if !matches!(room, RoomListEntry::Filled(_) | RoomListEntry::Invalidated(_)) {
            return false;
        }

        if let Some(state) = (self.state)(room) {
            state == RoomState::Invited
        } else {
            false
        }
    }
}

/// Create a new filter that will accept all filled or invalidated entries, but
/// filters out rooms that are not invites (see
/// [`matrix_sdk_base::RoomState::Invited`]).
pub fn new_filter(client: &Client) -> impl Filter {
    let client = client.clone();

    let matcher = InviteRoomMatcher {
        state: move |room| {
            let room_id = room.as_room_id()?;
            let room = client.get_room(room_id)?;
            Some(room.state())
        },
    };

    move |room_list_entry| -> bool { matcher.matches(room_list_entry) }
}
```

I think consistency is preferable here, even if your code is correct!